crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is.md

```diff
@@ -0,0 +1,26 @@
+# Narrowing for `is` conditionals
+
+## `is None`
+
+```py
+x = None if flag else 1
```

Comment:
I'm a little nervous that these tests will break once we add more type checking (`flag` is not defined here). I know it's a bit more boilerplate, but would you be able to do something like this in the tests?

```suggestion
def returns_bool() -> bool:
    return True

x = None if returns_bool() else 1
```

I _would_ suggest simply doing `x = None if bool() else 1`, but unfortunately I think that will also be prone to future breakage. We (accurately) infer `bool()` as returning `Literal[False]` rather than a random instance of `bool`, so in the future we'll probably be able to statically determine that one of the two branches is always taken, which would mean we wouldn't infer a union type

Reply:
That's actually something I noted down and wanted to ask. I just saw `flag` being used undeclared in other tests and copied that, but I also think it's not great. I actually thought about doing something like `… if random.randint(0, 1) else …`, but a better way would probably be to move all of this into a function, where the type checker can not assume anything about the value of the parameter:
```py
def f(x: Literal[1] | None):
    …
```

---

This seems like a common pattern in inference tests, so maybe we can come up with a better/shorter way to write them. I'm thinking something like
```py
x: Literal[1] | None = conjure()
```
where `conjure` would be a special function similar to Rusts `todo!()` — usually called `absurd` in type theory — of (hopefully not illegal) signature:
```py
def conjure[T]() -> T:
    ???
```
that we would somehow inject into those test examples. Not sure if that makes sense.

Reply:
> a better way would probably be to move all of this into a function, where the type checker can not assume anything about the value of the parameter:

Yes, this would definitely be my preference! Unfortunately we're not yet in a place where we can do that (we need to implement #13693 first, which shouldn't be _too_ tricky). We currently infer types from return annotations but not from parameter annotations.

> This seems like a common pattern in inference tests, so maybe we can come up with a better/shorter way to write them. I'm thinking something like
> 
> ```python
> x: Literal[1] | None = conjure()
> ```

Hrm, maybe... there's also advantages to keeping the test snippets as close to real-world Python as possible, though. The wonderful thing about having tests written in Markdown is that they double as living documentation of what our current inference capabilities are, and it's great for other readers of our tests if that documentation is as clear as possible.

Reply:
> there's also advantages to keeping the test snippets as close to real-world Python as possible, though

True. But I would claim that those `x = None if flag else 1` lines are not ideal in terms of their self-documenting property either. Definitely took me a moment to understand what's going on when I first saw them. And we have the problem with defining `flag`. What if we had something like the following pre-defined (never mind the actual implementation), possibly extended to a variadic function, is possible:
```py
def one_of[X, Y](x: X, y: Y) -> X | Y:
    if random.randint(0, 1):
        return x
    else:
        return y
```
This could be used like
```py
x = one_of(1, None)

reveal_type(x)  # Literal[1] | None
```

Reply:
I'm definitely open to it! Maybe you could create a followup issue on this for discussion? (I'd definitely want @carljm's input as well, as the chief author of the test framework 😄)

> (I understand that this might not yet be possible with the current state of red knot)

yes, generic functions are definitely a little way off for now 😄

Reply:
> Maybe you could create a followup issue on this for discussion

will do